TM1py/Services/CellService.py

```diff
@@ -179,6 +179,111 @@ def execute_mdx(self, mdx, cell_properties=None, top=None):
         finally:
             self.delete_cellset(cellset_id=cellset_id)
 
+    def get_data(self, mdx, format=None, cell_properites=None, dim_properties=None, top=None):
+        """ Execute an MDX query and return the results in the format specified
+
+        :param mdx_or_view: A string that is an MDX Query or view name
+        :param format: String specifying the format to return. Supported: 'celldict' (default), 'csv', 'raw', 'dataframe' (pandas), 'array', dygraph'
+        :param cell_properties: List of properties to be queried from the cell. E.g. ['Value', 'Ordinal', 'RuleDerived', ...]
+        :param dim_properties: List of properties to be queried from the dimension. E.g. ['UniqueName','Attributes', ...]
+        :param top: Integer limiting the number of cells and the number or rows returned
+        :return: Content in format specified above.
+        """
+        cellset_id = self.create_cellset(mdx=mdx)
+        try:
+            return self.extract_cellset(cellset_id=cellset_id, format=format, cell_properties=cell_properites, dim_properties=dim_properties, top=top)
+        finally:
+            self.delete_cellset(cellset_id=cellset_id)
+
+    def extract_cellset(self, cellset_id, format=None, cell_properties=None, dim_properties=None, top=None):
+        """ Extract data from an existing cellset and return the results in the format specified
+
+        :param cellset_id: String; ID of existing cellset
+        :param format: String specifying the format to return. Supported: 'celldict' (default), 'csv', 'raw', 'dataframe' (pandas), 'array', dygraph'
+        :param cell_properties: List of properties to be queried from the cell. E.g. ['Value', 'Ordinal', 'RuleDerived', ...]
+        :param dim_properties: List of properties to be queried from the dimension. E.g. ['UniqueName','Attributes', ...]
+        :param top: Integer limiting the number of cells and the number or rows returned
+        :return: Content in format specified above.
+        """        
+        if format == 'values':
+            data =  self.extract_cellset_values(cellset_id=cellset_id)
+            return (cell["Value"] for cell in data["Cells"])
+        elif format == 'csv':
+            return self.extract_cellset_csv(cellset_id=cellset_id)
+        elif format == 'dataframe':
+            raw_csv = self.extract_cellset_csv(cellset_id=cellset_id)
+            memory_file = StringIO(raw_csv)
+            return pd.read_csv(memory_file, sep=',')
+        else:
+            if not cell_properties:
```

Comment:
I think lines 218 to 221 are obsolete

Reply:
Oh, I only added these here since build_content_from_cellset needs the cell_properties to know which properties to add. It could be derived from the data and then it does not need them. Also, not sure if top is needed there since the query would be limiting the data.

Reply:
These lines are only needed to make sure the Utils.build_content_from_cellset method has the 'Ordinal' passed in the cell_properties (actually, to make sure it matches what the query in extract_cellset_full used.) I think Utils.build_content_from_cellset could derive that from the data and it does not need to be passed. I'm not sure it needs the top argument either, since the query already limited these values. What do you think about changing that function to take only raw_cellset_as_dict and value_precision, like the two new ones?

Reply:
Sorry, didn't mean to send duplicate comment.

Reply:
You are absolutely right. There is no need to query the Ordinal property from the cell. This needs to be fixed. 

Yes. It doesn’t necessarily need the top argument. We added it, so that theoretically the utility function is independent from the REST call. 
I think it makes sense, to add the value-precision as an argument. 

I wouldn’t want to remove the cell_properties argument. Sometimes it's useful to query other properties from the cell, like RuleDerived or CellAnnotation.